### DISCUSSION TITLE
TypeScript query builder expression API

### DISCUSSION THREAD

**Discussion by 1st1** [geldata/gel#3284] [General]
Let’s discuss the different design choices we have for the TypeScript query builder expression building API.

## Goals
It goes without saying that it isn’t possible to re-implement complete EdgeQL syntax in a programming language such as JavaScript or TypeScript. That said we should strive to get the new query builder API as close to EdgeQL as possible keeping the following goals in mind:

(1) The API should be composable.

(2) The API shouldn’t rename core EdgeQL operators / functions / concepts.

(3) The API should feel “intuitive” to a person who is familiar with EdgeQL. And vice-versa.

(4) The API should be convenient to use, and code that uses the API to build queries of various complexity should be readable.

## Functional API
This is what we have currently implemented.

An example EdgeQL expression `Movie.title ILIKE ‘blade runner%’ AND Movie.year >= 2000` can be represented with the following JS code:

```javascript
e.select(e.Movie, m => {
  title,
  filter: e.and(
    e.ilike(m.title, ‘blade runner%’), 
    e.gte(m.year, 2000)
  )
})
```

This is a standard way of implementing expression-building APIs in languages without operator overloading. And even in languages that do have operator overloading, it’s not usually possible to define entirely custom operators like “ILIKE”.

Pros: simple and somewhat common approach.

Cons: 

* users will have to learn the names of all EdgeQL operators in the API. For some like `DISTINCT` or `ILIKE` the mapping is obvious. For the infix `-` operator it could be anything: `minus`, `sub`, `subtract`; for the unary `-` operator it could be `minus`, `uminus`, `unaryMinus` , or even `negate`.
* there’s some mental effort required to read expressions built with this API, e.g. realize that `e.add(1, 2)` is `1 + 2`.

Both cons aren’t significant.

## Fluent API
An example EdgeQL expression `Movie.title ILIKE ‘blade runner%’ AND Movie.year >= 2000` can be represented with the following JS code:

```javascript
e.select(e.Movie, m => {
  title,
  filter: e.and(
    m.title.ilike(‘blade runner%’), 
    m.year.gte(2000)
  )
})
```

which can be even further simplified because methods are chainable:

```javascript
e.select(e.Movie, m => {
  title,
  filter: m.title.ilike(‘bladerunner%’).and(m.year.gte(2000))
})
```

The design can get to the extreme of also allowing to use stdlib functions as methods. E.g. the `str_upper(e.Movie.title) > ‘M’` EdgeQL expression would be represented as `e.Movie.title.str_upper().gt(‘M’)` with such API.

Pros: compact.

Cons:

* Many operators like ‘AND’ or `+` would have to have both “method” and “functional” versions to make it possible to group things differently.
* Departs from EdgeQL too far; there’s no notion of methods in EdgeQL.
* Really goes agains goals (2) and (3).

Overall, this approach doesn’t seem to offer much over the current functional API.

## Array-based API
An example EdgeQL expression `Movie.title ILIKE ‘blade runner%’ AND Movie.year >= 2000` can be represented with the following JS code:

```javascript
e.select(e.Movie, m => {
  title,
  filter: e.op([
    [m.title, 'ilike', ‘blade runner%’], 
    'and',
    [m.year, '>=', 2000]
  ])
})
```

Essentially, the API boils down to a function called `e.op()` that can accept arbitrarily nested JS tuple/triple arrays.

Pros: the API closely resembles EdgeQL: operands and operators have the same order and spelling, `[` act as parentheses.

Cons:

* I’m not aware of any other query builders implementing this design. It does look somewhat alien.
* I’m not sure we can implement proper error reporting for deeply nested arrays. Imagine that there’s a type error somewhere deep in the structure passed to `e.op()`; we’d need TypeScript compiler to pinpoint the exact element of the exact array literal where the error has been made. It’s an open question if that can be implemented.
* Not really a “con” as all other API versions suffer from this: Simple EdgeQL queries like `1 + 2 + 3` would require extra `[`s in this API, as in `[1, ‘+’, [2, ‘+’, 3]]`. We can only support 2- and 3- element arrays, not N-element arrays for this API.

## Hybrid functional API
An example EdgeQL expression `Movie.title ILIKE ‘blade runner%’ AND Movie.year >= 2000` can be represented with the following JS code:

```javascript
e.select(e.Movie, m => {
  title,
  filter: e.op(
    e.op(m.title, 'ilike', ‘blade runner%’), 
    'and',
    e.op(m.year, '>=', 2000)
  )
})
```

This is very close to the Array-based API defined in the previous section, with the only difference that we always use `e.op()` for every operator.

Pros: `e.op()`  is basically parentheses in EdgeQL; all operators are spelled exactly as in EdgeQL. Unlike the “Array-based API” this should be implementable with decent error reporting.

Cons: just like with “Array-based API”: this design is a bit unusual.

**Comment by 1st1** [comment]
I like the "Hybrid functional API" proposal. I think that the order of operands/operators combined with the ability of just spelling EdgeQL operators as-is is a great dx. IMO it isn't much different from the current functional API we have.

**Reply by elprans** [reply]
Same.  I think the "Hybrid functional" approach strikes the best balance between implementability, expressive power, and readability.  Ultimately, we are working around the absence of operator overloading in JavaScript, so replacing a hypothetical overloaded `m.title >= 2000` with `e.op(m.title, '>=', 2000)` makes sense and retains the shape of expressions.

**Reply by vpetrovykh** [reply]
"Hybrid functional" seems like a good balance, especially if implementing errors for deeply nested structures using `[]` is hard or impossible.

**Comment by spa5k** [comment]
Hybrid API looks much better, also edgedb query builder can take examples from Prisma on how to design it and make it a bit like that.

Also rather than doing 

```ts
e.select(e.Movie, m => {
  title,
  filter: e.op(
    e.op(m.title, 'ilike', ‘blade runner%’), 
    'and',
    e.op(m.year, '>=', 2000)
  )
})
```

Maybe it can be something like this - 

```ts
e.movie.select({
  title: true,
  filter: (
    (title, 'ilike', ‘blade runner%’), 
    'and',
    (year, '>=', 2000)
  )
})
```

**Reply by jaclarke** [reply]
Are those `()` in the example meant to be `[]`, and are `title` and `year` meant to be strings? Because as written I don't think that example is valid js. Also with `'title'` and `'year'` being strings, it could be ambiguous whether you're referring to the property name or if it's just a string literal.

**Reply by spa5k** [reply]
oh, i don't know if it's a correct js or not, but my main point was that we can ignore few of the repeated things like keeping e.op everywhere, same for m.year etc. and they can be referenced directly.

**Comment by kafein** [comment]
I think implementing LSP for EdgeDB queries will be a much better option. And it can be also used by other programming languages. After having auto-completion for the queries second step can be some kind of type generator based on queries. (Like Prisma)

I am sorry if it's far from the subject.

**Reply by elprans** [reply]
LSP and language-native query builders are largely orthogonal and we are working on both.  The JavaScript/TypeScript query builder is actually nearly complete (including schema reflection, codegen, CLI etc), and will have the first beta release soon.  This discussion is essentially to finalize the approach we take to implementing operator calls.

**Comment by black7375** [comment]
A library worth checking out: https://github.com/samchon/safe-typeorm
The auto-completion and query tuning are impressive.

**Reply by spa5k** [reply]
In same lines - https://github.com/koskimas/kysely

is also very impressive.
